## Diabetes Prediction On the basis of Different Test Results(insulin etc) Using Machine Learning Techniques

## Importing relevant libraries

In [1]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn import svm
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

## Loading the Data

In [2]:
df = pd.read_csv(r'D:\Data Science Project\8 -Diabetes\diabetes.csv')

In [3]:
df.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

## Checking For Null Values

In [4]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

## Balancing The Dataset

In [5]:
diabetic = df[df.Outcome == 1]
non_diabetic = df[df.Outcome == 0]

In [6]:
diabetic.shape

(268, 9)

In [7]:
non_diabetic.shape

(500, 9)

## Choosing as many non diabetics as there are diabetics in this data

In [8]:
sample_non_dibt = non_diabetic.sample(n=diabetic.shape[0])

In [10]:
print(diabetic.shape,sample_non_dibt.shape)

(268, 9) (268, 9)


In [11]:
df_new = pd.concat([diabetic,sample_non_dibt],axis = 0)

In [13]:
df_new.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
2            8      183             64              0        0  23.3   
4            0      137             40             35      168  43.1   
6            3       78             50             32       88  31.0   
8            2      197             70             45      543  30.5   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
2                     0.672   32        1  
4                     2.288   33        1  
6                     0.248   26        1  
8                     0.158   53        1

In [14]:
X = df_new.drop(columns = 'Outcome',axis = 1)
y = df_new['Outcome']

## Standardize The Input

In [15]:
scaler = StandardScaler()

In [16]:
scaler.fit(X)

StandardScaler()

In [17]:
std_dt = scaler.transform(X)

In [18]:
std_dt

array([[ 5.89145955e-01,  6.68923665e-01,  1.41007014e-01, ...,
         1.07152624e-01,  3.79188450e-01,  1.38814118e+00],
       [ 1.17939312e+00,  1.71545923e+00, -2.63703881e-01, ...,
        -1.23579449e+00,  5.09943088e-01, -1.54095820e-01],
       [-1.18159553e+00,  3.40012488e-01, -1.47783657e+00, ...,
         1.34579317e+00,  5.20548742e+00, -6.84159865e-02],
       ...,
       [-5.91348370e-01, -4.67314948e-01,  2.42184738e-01, ...,
        -4.93072345e-02,  5.85490213e-01, -5.82494987e-01],
       [-1.10120739e-03, -3.47710884e-01, -2.13115019e-01, ...,
        -1.41833099e+00, -1.87414981e-01,  2.74303348e-01],
       [-2.96224789e-01, -2.87908851e-01, -3.50139104e+00, ...,
        -1.20971785e+00, -8.99301344e-01, -9.25214321e-01]])

In [19]:
X = std_dt

## Training And Testing

In [21]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,stratify = y,random_state = 2)

In [22]:
classifier = svm.SVC(kernel = 'linear')
classifier.fit(X_train,y_train)

SVC(kernel='linear')

## Predicting the training data and checking for accuracy

In [24]:
X_train_prediction = classifier.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction,y_train)
training_data_accuracy

0.7827102803738317

## Making Prediction on the testing data and checking its accuracy

In [25]:
X_test_prediction = classifier.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction,y_test)
test_data_accuracy

0.6574074074074074

## Using RandomForestClassifier

In [26]:
model = RandomForestClassifier()

In [27]:
model.fit(X_train,y_train)

RandomForestClassifier()

## Predicting the training data and checking for accuracy

In [28]:
X_train_prediction = model.predict(X_train)
train_data_accuracy = accuracy_score(X_train_prediction,y_train)
train_data_accuracy

1.0

## Making Prediction on the testing data and checking its accuracy

In [29]:
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction,y_test)
test_data_accuracy

0.7314814814814815

## We had very little data thats why our model's accuracy is only 73%